RationaleMCP/0031/source-locations.md

```diff
@@ -0,0 +1,123 @@
+
+Each _source location decoration_ (or simply _decoration_ in this context) attaches some kind of source location information to a specific construct in the Flat Modelica code.
+
+Design decisions include:
+- Details of what a source location consists of: standardized or open for tool-specific choices?
+- Syntax of information included directly in a decoration.
+- Syntax for how a decoration is attached to a Flat Modelica construct.
+- How to handle need for (optional) big tables of verbose source location data.
+
+
+## Design goals
+
+Some design goals to keep in mind when comparing the design alternatives:
+- The detailed format of a source location needs to be flexible to serve the needs of different tools.
+- A tool should be able to ignore source locations originating from other tools.
+- Source locations may be needed on many levels of the Flat Modelica code, including component declarations, equations, modifications, and expressions.
+- Reduced Flat Modelica soruce code human readability is a concern when soruce locations are added.
+- Simple compression of the source locations could be handy.
+
+
+
+## Tool-specific table of source location details
+
+The Flat Modelica package level annotation may have a `SourceLocations`-annotation.
+There is no standardized content for this annotation, so it is currently only possible to use with vendor-specific annotations inside.
+For example:
+```
+package 'MyModel'
+model 'MyModel'
+…
+end 'MyModel';
+annotation(SourceLocations(__Wolfram(sourceLocationsFile = "MyModel.loc")));
+end 'MyModel';
+```
+
+As an alternative to pointing to a vendor-specific external file, it is also possible to include information directly in the annotation:
+```
+annotation(SourceLocations(__OpenModelica(
+  table = {
+    SourceLocation(file = "MoModel.mo", start = {1, 2}, end = {1, 7}),
+    SourceLocation(…),
+    …
+  }));
+```
+
+In the future, there could be standardized contents for the `SourceLocations` that all Flat Modelica tools are expected to recognize.
+
+
+## Syntax of information present in a source location decoration
+
+### Design alternatives
+
+Alternatives include:
+- Just an integer to be used with a separate table of details provided by `SourceLocations`-annotation.
+- Predefined record to avoid need for tool-specific table. (Details to be designed).
+- Use `annotation(SourceLocation(…))` to avoid need for attachment operator.
+
+The alternatives above could also be combined in different ways:
+- Number of supported formats:
+  - Standardize on just one of the alternatives.
+  - Allow more than one of the alternatives.
+- Number of informations included in a single source location decoration
+  - Just a single piece of information.
+  - Multiple pieces of information (possibly in different formats).
+
+### Conclusion
+
+None yet, but all discussion so far has pointed in the direction of just a single integer.
```

Comment:
According to today's decision above:
```suggestion
For now, only allow a single integer.  It would be easy to extend this in future versions of Flat Modelica.
```